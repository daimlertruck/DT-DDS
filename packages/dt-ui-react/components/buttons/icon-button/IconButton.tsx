import { BaseProps, Colors } from '../../../types';

import { IconButtonStyled } from './IconButton.styled';

export interface IconButtonProps extends BaseProps {
  onClick?: (e: React.MouseEvent<HTMLButtonElement>) => void;
  onMouseDown?: (e: React.MouseEvent<HTMLButtonElement>) => void;
  onMouseUp?: (e: React.MouseEvent<HTMLButtonElement>) => void;
  onMouseOver?: (e: React.MouseEvent<HTMLButtonElement>) => void;
  onMouseEnter?: (e: React.MouseEvent<HTMLButtonElement>) => void;
  color?: Colors | 'default';
  isDisabled?: boolean;
  ariaLabel?: string;
}

export const IconButton = ({
  children,
  dataTestId,
  isDisabled,
  ariaLabel,
  onClick,
  onMouseDown,
  onMouseUp,
  onMouseOver,
  onMouseEnter,
  color = 'primary',
}: IconButtonProps) => {
  const handleClick = (e: React.MouseEvent<HTMLButtonElement>) => {
    e.preventDefault();
    e.stopPropagation();
    onClick && onClick(e);
  };

  return (
    <IconButtonStyled
      aria-label={ariaLabel}
      color={color}
      data-testid={dataTestId ? dataTestId : 'icon-button'}
      disabled={isDisabled}
      onClick={handleClick}
      onMouseDown={onMouseDown}
      onMouseEnter={onMouseEnter}
      onMouseOver={onMouseOver}
      onMouseUp={onMouseUp}
    >
      {children}
    </IconButtonStyled>
  );
};
