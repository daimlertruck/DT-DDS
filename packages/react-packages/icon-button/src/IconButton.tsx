import { BaseProps, ComponentSize } from '@dt-ui/react-core';
import { ComponentPropsWithoutRef } from 'react';

import { IconButtonVariant } from './constants';
import { IconButtonStyled } from './IconButton.styled';

export interface IconButtonProps
  extends ComponentPropsWithoutRef<'button'>,
    BaseProps {
  onClick?: (e: React.MouseEvent<HTMLButtonElement>) => void;
  isDisabled?: boolean;
  ariaLabel?: string;
  variant?: IconButtonVariant;
  size?: ComponentSize;
}

export const IconButton = ({
  children,
  dataTestId,
  isDisabled,
  ariaLabel,
  variant,
  size,
  onClick,
  ...props
}: IconButtonProps) => {
  const handleClick = (e: React.MouseEvent<HTMLButtonElement>) => {
    if (isDisabled) {
      return;
    }

    e.preventDefault();
    e.stopPropagation();
    onClick && onClick(e);
  };

  return (
    <IconButtonStyled
      aria-label={ariaLabel}
      data-testid={dataTestId ? dataTestId : 'icon-button'}
      disabled={isDisabled}
      onClick={handleClick}
      size={size}
      variant={variant}
      {...props}
    >
      {children}
    </IconButtonStyled>
  );
};
