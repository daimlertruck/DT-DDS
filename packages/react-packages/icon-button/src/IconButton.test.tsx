import { withProviders } from '@dt-ui/react-core';
import { Icon } from '@dt-ui/react-icon';
import { fireEvent, render, screen } from '@testing-library/react';

import { IconButton, IconButtonProps } from './IconButton';

const handleClickMock = jest.fn();

describe('<IconButton /> component', () => {
  const ProvidedIconButton = withProviders(IconButton);

  const renderComponent = (props?: IconButtonProps) => {
    const { container } = render(
      <ProvidedIconButton {...props} onClick={handleClickMock}>
        <Icon code='edit' dataTestId='icon' />
      </ProvidedIconButton>
    );

    return {
      button: screen.getByTestId('icon-button'),
      icon: screen.getByTestId('icon'),
      container,
    };
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders with custom aria-label', () => {
    renderComponent({ ariaLabel: 'edit button' });

    const button = screen.getByLabelText('edit button');

    expect(button).toBeVisible();
  });

  it('calls onClick when clicked', () => {
    const { button } = renderComponent();

    expect(button).toHaveStyle('cursor: pointer');

    fireEvent.click(button);

    expect(handleClickMock).toHaveBeenCalledTimes(1);
  });

  it('does not call onClick if disabled and has cursor not-allowed', () => {
    const { button } = renderComponent({ isDisabled: true });

    expect(button).toHaveStyle('cursor: not-allowed');

    fireEvent.click(button);

    expect(handleClickMock).toHaveBeenCalledTimes(0);
  });

  it.each`
    size             | fontSize
    ${'extra-small'} | ${12}
    ${'small'}       | ${16}
    ${'medium'}      | ${20}
    ${'large'}       | ${24}
    ${'extra-large'} | ${32}
  `(
    'should have correct font size when size is $size',
    ({ size, fontSize }) => {
      const { icon } = renderComponent({ size });

      expect(icon).toHaveStyle({ fontSize: `${fontSize}px` });
    }
  );

  it.each`
    variant       | isDisabled
    ${'default'}  | ${false}
    ${'default'}  | ${true}
    ${'contrast'} | ${false}
    ${'contrast'} | ${true}
  `(
    'should have style if variant is $variant and isDisabled is $isDisabled',
    ({ variant, color, isDisabled }) => {
      const { container } = renderComponent({
        variant,
        color,
        isDisabled,
      });

      expect(container).toMatchSnapshot();
    }
  );
});
