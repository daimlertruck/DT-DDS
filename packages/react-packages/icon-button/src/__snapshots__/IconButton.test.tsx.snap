// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IconButton /> component should have style if variant is contrast and isDisabled is false 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: none;
  background-color: transparent;
  cursor: pointer;
}

.emotion-0 i {
  font-size: 24px;
  color: #ffffff;
}

.emotion-0:not(:disabled)>i:hover,
.emotion-0:not(:disabled)>i:active {
  color: #73aec2;
}

.emotion-0:focus-visible {
  outline: 2px solid #000000;
}

.emotion-2 {
  font-family: DTUI-icons-outlined;
  font-weight: normal;
  font-style: normal;
  font-size: 24px;
  display: inline-block;
  line-height: 1;
  text-transform: none;
  letter-spacing: normal;
  word-wrap: normal;
  white-space: nowrap;
  font-variation-settings: 'FILL' 0;
  color: #292929;
  -webkit-font-smoothing: antialiased;
  text-rendering: optimizeLegibility;
  -moz-osx-font-smoothing: grayscale;
  font-feature-settings: 'liga';
}

<div>
  <button
    class="emotion-0 emotion-1"
    data-testid="icon-button"
  >
    <i
      class="emotion-2"
      color="#292929"
      data-testid="icon"
    >
      edit
    </i>
  </button>
</div>
`;

exports[`<IconButton /> component should have style if variant is contrast and isDisabled is true 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: none;
  background-color: transparent;
  cursor: not-allowed;
}

.emotion-0 i {
  font-size: 24px;
  color: #292929;
}

.emotion-0:not(:disabled)>i:hover,
.emotion-0:not(:disabled)>i:active {
  color: #73aec2;
}

.emotion-0:focus-visible {
  outline: 2px solid #000000;
}

.emotion-2 {
  font-family: DTUI-icons-outlined;
  font-weight: normal;
  font-style: normal;
  font-size: 24px;
  display: inline-block;
  line-height: 1;
  text-transform: none;
  letter-spacing: normal;
  word-wrap: normal;
  white-space: nowrap;
  font-variation-settings: 'FILL' 0;
  color: #292929;
  -webkit-font-smoothing: antialiased;
  text-rendering: optimizeLegibility;
  -moz-osx-font-smoothing: grayscale;
  font-feature-settings: 'liga';
}

<div>
  <button
    class="emotion-0 emotion-1"
    data-testid="icon-button"
    disabled=""
  >
    <i
      class="emotion-2"
      color="#292929"
      data-testid="icon"
    >
      edit
    </i>
  </button>
</div>
`;

exports[`<IconButton /> component should have style if variant is default and isDisabled is false 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: none;
  background-color: transparent;
  cursor: pointer;
}

.emotion-0 i {
  font-size: 24px;
  color: #292929;
}

.emotion-0:not(:disabled)>i:hover,
.emotion-0:not(:disabled)>i:active {
  color: #00677f;
}

.emotion-0:focus-visible {
  outline: 2px solid #000000;
}

.emotion-2 {
  font-family: DTUI-icons-outlined;
  font-weight: normal;
  font-style: normal;
  font-size: 24px;
  display: inline-block;
  line-height: 1;
  text-transform: none;
  letter-spacing: normal;
  word-wrap: normal;
  white-space: nowrap;
  font-variation-settings: 'FILL' 0;
  color: #292929;
  -webkit-font-smoothing: antialiased;
  text-rendering: optimizeLegibility;
  -moz-osx-font-smoothing: grayscale;
  font-feature-settings: 'liga';
}

<div>
  <button
    class="emotion-0 emotion-1"
    data-testid="icon-button"
  >
    <i
      class="emotion-2"
      color="#292929"
      data-testid="icon"
    >
      edit
    </i>
  </button>
</div>
`;

exports[`<IconButton /> component should have style if variant is default and isDisabled is true 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: none;
  background-color: transparent;
  cursor: not-allowed;
}

.emotion-0 i {
  font-size: 24px;
  color: #a3a3a3;
}

.emotion-0:not(:disabled)>i:hover,
.emotion-0:not(:disabled)>i:active {
  color: #00677f;
}

.emotion-0:focus-visible {
  outline: 2px solid #000000;
}

.emotion-2 {
  font-family: DTUI-icons-outlined;
  font-weight: normal;
  font-style: normal;
  font-size: 24px;
  display: inline-block;
  line-height: 1;
  text-transform: none;
  letter-spacing: normal;
  word-wrap: normal;
  white-space: nowrap;
  font-variation-settings: 'FILL' 0;
  color: #292929;
  -webkit-font-smoothing: antialiased;
  text-rendering: optimizeLegibility;
  -moz-osx-font-smoothing: grayscale;
  font-feature-settings: 'liga';
}

<div>
  <button
    class="emotion-0 emotion-1"
    data-testid="icon-button"
    disabled=""
  >
    <i
      class="emotion-2"
      color="#292929"
      data-testid="icon"
    >
      edit
    </i>
  </button>
</div>
`;
